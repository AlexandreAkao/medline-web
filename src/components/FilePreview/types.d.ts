type FilePreviewStatusTypes = 'ready' | 'waiting' | 'cancel';

interface IFilePreviewProps {
  children: React.ReactNode;
  status: FilePreviewStatusTypes;
}
