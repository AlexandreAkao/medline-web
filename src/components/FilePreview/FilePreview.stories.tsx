import { ComponentStory, ComponentMeta } from '@storybook/react';

import FilePreview from 'components/FilePreview';
import colors from 'styles/colors';

export default {
  title: 'FilePreview',
  component: FilePreview,
  decorators: [
    Story => (
      <div
        style={{
          background: colors.primary,
          width: '100%',
          height: '100%',
          padding: '50px',
        }}
      >
        <Story />
      </div>
    ),
  ],
} as ComponentMeta<typeof FilePreview>;

const Template: ComponentStory<typeof FilePreview> = args => <FilePreview {...args} />;

export const Primary = Template.bind({});
Primary.args = {
  children: 'Primary - Description - 99/99/9999',
};
