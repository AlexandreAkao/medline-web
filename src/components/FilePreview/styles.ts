import styled from 'styled-components';

import colors from 'styles/colors';

export const FilePreviewContainer = styled.div`
  background-color: ${colors.white.normal};
  box-shadow: 0px 0px 20px 0px rgba(0, 0, 0, 0.2);
  padding: 20px;
  border-radius: 12px;
  display: flex;
  align-items: center;
  justify-content: space-between;
  width: 100%;
`;

export const FilePreviewInfo = styled.span`
  font-size: 1.125rem;
  font-weight: 600;
`;

export const FilePreviewActionContainer = styled.div`
  .icon-action {
    margin: 0 0.5rem;
  }
`;
