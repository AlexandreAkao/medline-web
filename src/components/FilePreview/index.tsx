import { HiDownload, HiCheck, HiX } from 'react-icons/hi';
import { FaHourglassHalf } from 'react-icons/fa';
import { useMemo } from 'react';

import { FilePreviewActionContainer, FilePreviewContainer, FilePreviewInfo } from 'components/FilePreview/styles';
import { downloadFile } from 'utils/file';
import api from 'service/api';
import colors from 'styles/colors';

function FilePreview({ children, status, file }: IFilePreviewProps) {
  const IconStatus = useMemo(() => {
    switch (status) {
      case 'ready':
        return HiCheck;
      case 'cancel':
        return HiX;
      case 'waiting':
        return FaHourglassHalf;
      default:
        return FaHourglassHalf;
    }
  }, [status]);

  const handleDownloadFile = async () => {
    if (file) {
      const data = await api.get(`file/${file?.id}`, { responseType: 'blob' });
      downloadFile(data.data, file?.name ?? '');
    }
  };

  return (
    <FilePreviewContainer>
      <FilePreviewInfo>{children}</FilePreviewInfo>
      <FilePreviewActionContainer>
        <IconStatus size={25} className="icon-action" data-testid={`icon-${status}-test`} />
        <HiDownload
          size={25}
          color={colors.primary.darkest}
          className="icon-action download"
          onClick={handleDownloadFile}
          data-testid="icon-download-test"
        />
      </FilePreviewActionContainer>
    </FilePreviewContainer>
  );
}

export default FilePreview;
