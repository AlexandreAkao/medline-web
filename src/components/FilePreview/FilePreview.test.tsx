import { render, screen } from '@testing-library/react';
import { composeStories } from '@storybook/testing-react';

import * as stories from 'components/FilePreview/FilePreview.stories';
import FilePreview from 'components/FilePreview';

const { Primary } = composeStories(stories);

describe('FilePreview', () => {
  const renderComponent = (customProps: Partial<IFilePreviewProps> = {}) => {
    const props: IFilePreviewProps = {
      children: 'FilePreview example',
      status: 'ready',
      file: null,
      ...customProps,
    };
    return render(<FilePreview {...props} />);
  };

  it('should match with snapshot', () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  it('should render a primary Button', () => {
    render(<Primary />);
    const containerlement = screen.getByText(/Primary - Description - 99\/99\/9999/i);
    expect(containerlement).not.toBeNull();
  });

  describe('should render icon status', () => {
    it('check icon when status is ready', () => {
      renderComponent();
      const containerlement = screen.getByText(/FilePreview example/i);
      const statusIconElement = screen.getByTestId('icon-ready-test');
      expect(containerlement).toBeInTheDocument();
      expect(statusIconElement).toBeInTheDocument();
    });

    it('hourglass icon when status is ready', () => {
      renderComponent({ status: 'waiting' });
      const containerlement = screen.getByText(/FilePreview example/i);
      const statusIconElement = screen.getByTestId('icon-waiting-test');
      expect(containerlement).toBeInTheDocument();
      expect(statusIconElement).toBeInTheDocument();
    });

    it('cancel icon when status is ready', () => {
      renderComponent({ status: 'cancel' });
      const containerlement = screen.getByText(/FilePreview example/i);
      const statusIconElement = screen.getByTestId('icon-cancel-test');
      expect(containerlement).toBeInTheDocument();
      expect(statusIconElement).toBeInTheDocument();
    });
  });
});
