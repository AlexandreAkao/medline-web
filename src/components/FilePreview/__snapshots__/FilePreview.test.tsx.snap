// Vitest Snapshot v1

exports[`FilePreview > should match with snapshot 1`] = `
<div>
  <div
    class="sc-bczRLJ iFSACV"
  >
    <span
      class="sc-gsnTZi jjMMDi"
    >
      FilePreview example
    </span>
    <div
      class="sc-dkzDqf bYyGGF"
    >
      <svg
        class="icon-action"
        data-testid="icon-ready-test"
        fill="currentColor"
        height="25"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 20 20"
        width="25"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M16.707 5.293a1 1 0 010 1.414l-8 8a1 1 0 01-1.414 0l-4-4a1 1 0 011.414-1.414L8 12.586l7.293-7.293a1 1 0 011.414 0z"
          fill-rule="evenodd"
        />
      </svg>
      <svg
        class="icon-action download"
        color="#257472"
        fill="currentColor"
        height="25"
        stroke="currentColor"
        stroke-width="0"
        style="color: rgb(37, 116, 114);"
        viewBox="0 0 20 20"
        width="25"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M3 17a1 1 0 011-1h12a1 1 0 110 2H4a1 1 0 01-1-1zm3.293-7.707a1 1 0 011.414 0L9 10.586V3a1 1 0 112 0v7.586l1.293-1.293a1 1 0 111.414 1.414l-3 3a1 1 0 01-1.414 0l-3-3a1 1 0 010-1.414z"
          fill-rule="evenodd"
        />
      </svg>
    </div>
  </div>
</div>
`;
